api/src/test/java/org/openmrs/api/impl/PatientServiceImplUnitTest.java

```diff
@@ -0,0 +1,1379 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.impl;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.openmrs.*;
+import org.openmrs.api.*;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.PatientDAO;
+import org.openmrs.messagesource.MessageSourceService;
+import org.openmrs.parameter.EncounterSearchCriteria;
+import org.openmrs.patient.IdentifierValidator;
+import org.openmrs.patient.impl.LuhnIdentifierValidator;
+import org.openmrs.patient.impl.VerhoeffIdentifierValidator;
+import org.openmrs.person.PersonMergeLog;
+import org.openmrs.person.PersonMergeLogData;
+import org.openmrs.util.OpenmrsUtil;
+import org.openmrs.validator.PatientIdentifierValidator;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PowerMockIgnore;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import java.util.*;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+import static org.mockito.MockitoAnnotations.initMocks;
+import static org.openmrs.util.OpenmrsConstants.GLOBAL_PROPERTY_DEFAULT_PATIENT_IDENTIFIER_VALIDATOR;
+import static org.powermock.api.mockito.PowerMockito.mockStatic;
+import static org.powermock.api.mockito.PowerMockito.verifyPrivate;
+
+@RunWith(PowerMockRunner.class)
```

Comment:
Did you try testing the PatientServiceImpl without the PowerMockRunner? The issue with this currently is that it does not show up in the test coverage. 

An example is
https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/api/AdministrationServiceUnitTest.java

Can you come up with an assertion without relying on `verifyPrivate`? So only call public methods and check the output or a referenced object that might have been mutated.

Reply:
it doesn't work without PowerMockRunner
Without using powermock i'll have to set up whole context (i tried setting up only the things that i needed, but that doesn't work)

I've added some comments to https://issues.openmrs.org/browse/TRUNK-5150..  do check them out... :) @teleivo

Reply:
@ayush2297 it does work at least for some of the tests. Lets please try to come up with smaller commits and do this step by step. I can for example add

```java
	@Test
	public void exitFromCare_shouldThrowApiExceptionIfDateExitedIsNull() {

		assertThrows(APIException.class,
			() -> patientService.exitFromCare(new Patient(), null, new Concept()));
	}
```

into https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/api/impl/PatientServiceImplTest.java just fine. Please create a commit with unit tests added to https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/api/impl/PatientServiceImplTest.java Once we have exhausted our options we can still resort to PowerMock.

ExpectedException is partially deprecated, see
https://github.com/junit-team/junit4/wiki/Exception-testing#expectedexception-rule
the favored approach seems to now be
https://github.com/junit-team/junit4/wiki/Exception-testing#using-assertthrows-method

Reply:
okay @teleivo.... thanks for the help... 
i changed my tests .. now instead of using PowerMock i am using BaseContextMockTest.. 
I'll merge the tests into PatientServiceImplTest.

waiting for the naming convention confirmation  before pushing the code.